### PR TITLE
HPCC-19052 Add epoll handler threads

### PR DESCRIPTION
**Pull request by mckellyln**
Signed-off-by: Mark Kelly <mark.kelly@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Full QA testing with config parameter epoll_hdlperthrd unset and set to various values.

### CODE REVIEW

File: system/jlib/jsocket.cpp [hunk 1]
```diff
@@ -5152,95 +5167,145 @@ class CSocketEpollThread: public CSocketBaseThread
...
+                    return;
+            }
+            else if (!(addrm&0x2))
+            {
+                if (threads.item(i).remove(sock))
+                    return;
+            }
+        }
+        if (addrm&0x1)
+            return;
```
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: Or could use previous method:
```
        bool added = false;
        ForEachItemIn(i,threads)
        {
            if (added)
            {
                if (threads.item(i).remove(sock))
                    break;
            }
            else
                added = threads.item(i).add(sock,mode,nfy);
        }
        if (added)
            return;
```

File: system/jlib/jsocket.cpp [hunk 2]
```diff
@@ -4914,23 +4917,33 @@ class CSocketEpollThread: public CSocketBaseThread
...
-            return false;
+            return 0;
         }
         CriticalBlock block(sect);
-        removeSock(sock);
+        if (terminating)
+            return 0;
+        unsigned rm = 0;
+        if (removeSock(sock))
+            rm = 0x2;
```
Comment (ghalliday) [inline system/jlib/jsocket.cpp]: This really should use flags from an enumeration or similar - otherwise it isn't clear what is being checked.

**Comment by ghalliday** [comment]
@mckellyln it looks like a good change.  I didn't see any problems, just one comment on readability of the mask using 0x1 and 0x2

**Comment by mckellyln** [comment]
@ghalliday ok, I'll add an enum for these bits.

**Comment by mckellyln** [comment]
@ghalliday added SOCK_ADDED, SOCK_REMOVED

**Pull request merged**